src/targets/gpu/device/include/migraphx/gpu/device/scan.hpp

```diff
@@ -52,6 +52,51 @@ __device__ void block_scan(index idx, Op op, T init, index_int n, Input input, O
                   output);
 }
 
+template <index_int N,
+          class Op,
+          class T,
+          class ForStride,
+          class Input,
+          class Output,
+          MIGRAPHX_REQUIRES(not std::is_integral<ForStride>{})>
+__device__ void
+reverse_block_scan(index idx, Op op, T init, ForStride fs, Input input, Output output)
+{
+    using type = decltype(input(deduce_for_stride(fs)));
+    MIGRAPHX_DEVICE_SHARED type buffer[N];
+    type x = init;
+    fs([&](auto i) {
+        if(idx.local == idx.nlocal() - 1)
+            buffer[idx.local] = op(input(i), x);
+        else
+            buffer[idx.local] = input(i);
+        __syncthreads();
+        for(index_int s = idx.nlocal() / 2; s >= 1; s /= 2)
+        {
+            if(idx.local + s < idx.nlocal())
+            {
+                buffer[idx.local + s] = op(buffer[idx.local], buffer[idx.local + s]);
+            }
+            __syncthreads();
+        }
+        x = buffer[idx.local];
+        output(i, buffer[idx.local]);
+    });
+}
```

Comment:
So a seperate implementation shouldn't be necessary, I think we can use a function combinator on the input and output to do this:

```cpp
template<class F>
constexpr auto reverse_scan(index_int n, F f)
{
    return [=](auto i, auto&&... xs) {
        return f(n-i-1, xs...);
    };
}
template <index_int N, class Op, class T, class Input, class Output>
__device__ void reverse_block_scan(index idx, Op op, T init, index_int n, Input input, Output output)
{
    block_scan<N>(idx,
                  op,
                  init,
                  n,
                  reverse_scan(input),
                  reverse_scan(output));
}
```

And `reverse_block_scan` may not be needed as you could do this directly in the `prefix_scan_sum` function. I think something similar could be done for exclusive scan as well, maybe something like:

```cpp
template<class F>
constexpr auto exclusive_scan(index_int n, F f)
{
    return [=](auto i, auto&&... xs) {
        auto j = i + 1;
        if (j < n)
            f(j, xs...);
    };
}
```

And apply this to the output.

Reply:
Made the changes for reverse_scan, but I haven't been able to get exclusive_scan to work. When I attempt to use it in the same way that I've applied reverse_scan, I get this error:
```
AMDMIGraphX/src/targets/gpu/device/include/migraphx/gpu/device/scan.hpp:23:39: error: array has incomplete element type 'type' (aka 'void')
    MIGRAPHX_DEVICE_SHARED type buffer[N];
                                      ^
```